(nexus-repository-migrator) Fixes Nexus 3.71 Breaking Change Upgrade

JPRuskin opened the pull request:
## Description

We now have a new package, `nexus-repository-migrator`, that contains logic to migrate the Nexus Repository database from 3.70.* to 3.71.*.

This new migration process:
- Checks to see if the version we're on can be migrated from (e.g. ==3.70)
- Checks to see if the install is using OrientDb (and so needs to be upgraded)
- Checks to see if the machine has sufficient memory and disk space
- Backs up the OrientDb files from the $DataDir\db directory
- Downloads the migration tool
- Migrates the backed up databases
- Copies the migrated file(s) back to the database directory
- Updates the Nexus properties to enable the datastore

By default, the package just downloads the migrator file and adds the `ConvertFrom-NexusOrientDb` script to PATH. You can also pass `/Migration` to run it during the package installation.

After this is completed, you should be able to upgrade using the next available version of Nexus Repository - the 3.71.0.6 version will _still_ break due to the merged files!

This works back to PowerShell v3, largely due to `Compress-ArchiveCompat`.

As mentioned, we now ensure that we are removing and replacing the program directory instead of copying over / merging it with the new content, as the upgrade to JRE and Nexus broke stuff with that strategy. The changes now source all connection variables in Wait-NexusAvailability (other than Hostname) from the on-disk configuration, and automatically backs-up and restores the port and SSL configuration if they are defined in the configuration. This should result in a smoother experience for people in future upgrades, and ensure `Wait-NexusAvailability` fails less frequently on SSL enabled installations.

Finally, we re-enable the update script for Nexus Repository.

The update script for the package will only find packages that match the 3.70 to 3.71 migration, currently! This is a mindful choice that we can reconsider in the future.

## Motivation and Context

Nexus' recent update has added some fairly catastrophic breaking changes, and
necessitated a large addition to the install script for the sake of migrating
the database to the new H2 format (previously only available to pro customers)

Fixes #2516

## How Has this Been Tested?

- Tested on Windows Server 2022 (PowerShell 5), upgrading from 3.70.x, with configured SSL
- Tested on Windows Server 2012 (PowerShell 3), upgrading from 3.70.x, without SSL

## Types of changes

- ~[ ] Bug fix (non-breaking change which fixes an issue)~
- ~[ ] New feature (non-breaking change which adds functionality)~
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- ~[ ] Migrated package (a package has been migrated from another repository)~

## Checklist:
- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] I have updated the package description and it is less than 4000 characters.
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)
- [ ] The added/modified package passed install/uninstall in the Chocolatey Test Environment(https://github.com/chocolatey-community/chocolatey-test-environment/). _Note that we don't support the use of any other environment_.
- [ ] The changes only affect a single package (not including meta package).

The changes affect two packages, but this has been agreed with other members of the community to be alright.

JPRuskin committed:
(nexus-repository) Adds Migration Logic for Version 3.71

Nexus' recent update has added some fairly catastrophic breaking changes, and
necessitated a large addition to the install script for the sake of migrating
the database to the new H2 format (previously only available to pro customers)

This new migration process:
- Checks to see if the version we're on can be migrated from (e.g. ==3.70)
- Checks to see if the install is using OrientDb (and so needs to be upgraded)
- Checks to see if the machine has sufficient memory and disk space
- Backs up the OrientDb files from the $DataDir\db directory
- Downloads the migration tool
- Migrates the backed up databases
- Copies the migrated file(s) back to the database directory
- Updates the Nexus properties to enable the datastore
- Deletes the entire target folder (there's a separate discussion that we
    may want to be doing this anyway)
- Continues with the standard upgrade process

This works back to PowerShell v3, largely due to `Compress-ArchiveCompat`

JPRuskin committed:
(nexus-repository) Improves Parameter Handling

Sources all connection variables in Wait-NexusAvailability (other than Hostname) from the on-disk configuration.

Also automatically backs-up and restores the port and SSL configuration if they are defined in the configuration.

JPRuskin commented:
I can strip out the second commit if we prefer.

steviecoaster commented on automatic/nexus-repository/nexus-repository.nuspec:
```diff
@@ -40,6 +40,12 @@ You can pass parameters as follows:
+- **ATTENTION BREAKING CHANGE FOR UPGRADES FROM VERSIONS BEFORE 3.71.0.6**  
+    Nexus has upgraded the bundled version of Java, and sunset the OrientDb database type.
+    This means that you will need to migrate your database from OrientDb, if you are using it.
+    The version after 3.71.0.6 will attempt to migrate the database during the package upgrade.
```
after 3.71.0.6? 

This reads like this version won't migrate to H2, when it in fact wil.

JPRuskin replied:
To be clear, version 3.71.0.6 will never have the upgrade logic in. :(

steviecoaster replied:
Ooooh right, as that's already hit the feed approved. Derp. Ignore me.

steviecoaster commented on automatic/nexus-repository/tools/chocolateyinstall.ps1:
```diff
@@ -47,6 +48,76 @@ if ($pp.ContainsKey("BackupSslConfig")) {
+} elseif ($CurrentlyInstalledVersion -and [version]($CurrentlyInstalledVersion -replace '-\d+$') -eq "3.70.1" -and (Test-NexusDatabaseType -Type "OrientDb" -DataDir $TargetDataFolder -ProgramDir $TargetFolder)) {
+  Write-Host "Preparing for database migration..."
+  $MigrationFiles = Join-Path $ExtractFolder "dbmigration"
+  if ((Get-PSDrive $MigrationFiles[0]).Free -lt ($RequiredSpace = [Math]::Max((3 * (Get-ChildItem -Path $TargetDataFolder\nexus3\db -Recurse | Measure-Object Length -Sum).Sum), 10GB))) {
```
We may want to break this down a bit so it's easier to follow what's goin on. It works, but it is a bit of a symbol salad right now.

JPRuskin replied:
I did think about pushing this all out into a Test-migrationrequired function. It would be moving the problem, but maybe good for those who read the installer scripts!

steviecoaster replied:
Someone still has to review the code though whether it is here directly or in helper. Pick a spot, and use more variable for readability. We don't have to pay for them :)

JPRuskin replied:
Ah, I misunderstood your complaint.

TheCakeIsNaOH commented on automatic/nexus-repository/tools/helpers.ps1:
```diff
@@ -105,3 +97,189 @@ function Wait-NexusAvailability {
+    foreach ($Database in $IncludedDatabases) {
+      Push-Location (Join-Path $DataDir "nexus3/db/$Database")
+      $ArchiveArgs = @{
+        Path = Get-ChildItem -Path (Join-Path $DataDir "nexus3/db/$Database/*") -Exclude "cache.stt" -File
```
My understanding is that to ensure the database in a consistent state, the Nexus task to export the database should be used. I'm just concerned that this could be creating a backup with the database not working correctly.

JPRuskin replied:
That's why I'm shutting down the service.

To be clear, I have both:
- Used BeyondCompare to check that my backups are file-identical (the only difference is the compression algo used)
- Tested a _bunch_ of times that migrations succeed, and import all of the data (that I can see) successfully

JPRuskin replied:
Resolving this off, as I hope I answered the question / addressed the concern. If not, please let me know.

TheCakeIsNaOH commented on automatic/nexus-repository/tools/chocolateyinstall.ps1:
```diff
@@ -35,18 +42,89 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
+  try {
+    Push-Location $MigrationFiles
+    Write-Host "Migrating database from 'OrientDb' to 'H2'"
+    $JavaPath = Join-Path $TargetFolder "jre/bin/java.exe"
```
Is this java binary always going to be java 8/11, i.e. a compatible version with the migrator `.jar`?

JPRuskin replied:
Yes, the bundled JRE version in the Chocolatey package is always, on versions <3.71, going to be 8/11.

TheCakeIsNaOH commented on automatic/nexus-repository/tools/chocolateyinstall.ps1:
```diff
@@ -35,18 +42,89 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
+  # See: https://help.sonatype.com/en/orientdb-downloads.html
+  $MigratorDownload = @{
+    PackageName = $env:ChocolateyPackageName
+    Url = 'https://download.sonatype.com/nexus/nxrm3-migrator/nexus-db-migrator-3.70.1-03.jar'
```
Is it worth breaking this migrator out into a separate package so it can be updated separately, since the download page mentions that download urls for older version(s) will be removed if/when it gets updated.

JPRuskin replied:
I'm going back and forth between "we should continue to support migration for several versions, so this makes good sense" (as we'd have to consider adding the migration tool to the update script in either case) and "unless we plan to support migrating indefinitely, and will consequently have to continue to upgrade this tool and test mirgations, maybe we should just say 'the upgrade path requires you to go to 3.70.1, then 3.71.x, and then it's all good'".

![image](https://github.com/user-attachments/assets/a7761921-629f-4c17-92b8-a8363f27006d)

I dislike this broken chain.

JPRuskin commented:
My one fun question is "do we want to handle trying to fix up installs where they have attempted to upgrade to 3.71.0.6 and it's failed"? 

I suspect they will be in a fun state.

JPRuskin committed:
(nexus-repository-migrator) Extracts nexus-repository Migration Logic

It was raised that we should have a separate package for the Nexus Repository Migrator logic.
This commit removes that code from the Nexus Repository package, and adds it to a separate package that can be installed.

steviecoaster commented on automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1:
```diff
@@ -0,0 +1,128 @@
+  [switch]$Force
+)
+
+Import-Module $PSScriptRoot\nexus-helpers.ps1 -Verbose:$false
```
Where is this file?

JPRuskin replied:
The nuspec pulls the helpers file from nexus repository. I had written it to use it as psm1, but it turned out that NuGet treats file extension changes as a folder instead.

JPRuskin committed:
(nexus-repository) Removes ProgramDir before Copying Files

The upgrade to JRE exposed that the package was simply copying new files over the old ones, which can break things when old files combine with new.

This change solidifies the package logic to backup SSL configuration if it's present, and specifically removes the program directory before re-copying the latest files.

This should ensure that Nexus can start, after the upgrade.

JPRuskin commented:
Just to keep people updated:

I'm (we're) happy with the code in this PR, so it's ready for review. However, we're keeping it in draft because Sonatype have a few changes they're making to the migrator to not break indexes and such. Until that new version is released, we [probably] shouldn't release this package. It's arguable.

That said, please do let loose with feedback.

JPRuskin committed:
(nexus-repository-migrator) Adds Longer Timeout to AttemptMigration

The post-migration launch process can take significantly longer than normal, due to the additional work Nexus is doing.

This change adds a configurable timeout to the Wait-NexusAvailability command, and uses it in the Nexus-Repository-Migrator package to increase the timeout from 3 minutes to 15.

Windos reviewed (commented):
Neither a solid approve or change request from me, just the result of my personal testing.

As mentioned in the inline comments, I originally didn't see that the migrator package had a parameter that needed to be passed in if I wanted to attempt the migration at install time. This led to issues with my trying to run the migrator after the fact.

I mentioned at least one variable defined in the install which is then used in the migration script (I didn't see any others) but even when manually re-creating that variable the migration did not succeed.

I ended up having to force a re-install of the migration package with the `/AttemptMigration` parameter and then the migration worked as expected.

I'm not sure if the best way to stop people like me from holding this wrong is in the documentation (i.e. adding to the "you need to install this migration package" error) or code (making attempting the migration the default behaviour), that is probably worth a discussion?

JPRuskin commented on automatic/nexus-repository/tools/chocolateyinstall.ps1:
```diff
@@ -34,12 +42,21 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
+if (Test-NexusMigratorRequired -DataDir $TargetDataFolder -ProgramDir $TargetFolder) {
+  Write-Error (@(
+    "This upgrade will fail if you do not migrate your database from OrientDb before proceeding."
+    "You can do this with the nexus-repository-migrator package, or following the upgrade instructions."
```
I was steered away from automating the actions too much in the last community package discussion, so I ended up defaulting to just providing the file(s). I've done an attempt at improving the messaging.

JPRuskin commented on automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1:
```diff
@@ -0,0 +1,113 @@
+        "-XX:+UseG1GC"
+        "-XX:MaxDirectMemorySize=28672M"
+        "-jar"
+        $MigratorDownload.FileFullPath
```
Embarrassing! I used to have a fully separate download block in this script, but removed it when I realised I'd have to keep two checksums up to date 😅 

Fixed this up, great catch, thank you!

JPRuskin committed:
(nexus-repository) Update Package and Feedback from PR

The existing package no longer works, so we need to update the link used and apply a patchfix version.

JPRuskin commented on automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec:
```diff
@@ -0,0 +1,79 @@
+
+This package supports the following parameters:
+
+* `/AttemptMigration` - Attempts to migrate your existing Sonatype Nexus install from OrientDb to a H2 database.
```
> That sounds like 'We-will-give-this-a-try-fingers-crossed'

I mean, it was written by me 😅 That said, it will very much _attempt_ - it'll check a bunch of stuff first, and fail before calling the migrator if (for instance) you have less than the recommended amount of memory or disk space. I'll adjust the wording, in any case, to sound more hopeful.

JPRuskin replied:
The parameter is now `/Migrate`?

pauby replied:
It is. The text still says 'attempt':

> `/Migrate` - Attempts to migrate your existing Sonatype Nexus install from OrientDb to a H2 database.

pauby reviewed (changes requested):
Note I'm not reviewing the code or testing the install / upgrade / uninstall.

pauby reviewed (dismissed):
I know I've said this a few times, but just being clear again, I have not checked the code or tested this package. Just the wording.

Windos commented:
I can re-do my testing of the code in the morning, should have results up ~16 hours

pauby commented:
@TheCakeIsNaOH as you reviewed this previously, are you able to review this again and approve or add comments?

Unless anybody else feels strongly, once we have @Windos and @TheCakeIsNaOH review, I suggest we merge this.

Windos commented on automatic/nexus-repository/tools/helpers.ps1:
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
-  }
-
-  $NexusUri = "$($nexusScheme)://$($hostname):$($nexusPort)$($nexusPath)"
+  $NexusUri = Get-NexusUri -Hostname $Hostname
```
```suggestion
  $NexusUri = Get-NexusUri -HostnameOverride $Hostname
```

Just wondering if this should be changed? Not sure if PowerShell is doing magic to make `-Hostname` match `HostnameOverride`

Windos commented on automatic/nexus-repository/tools/helpers.ps1:
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
 }
 
-function Wait-NexusAvailability {
+function Get-NexusUri {
```
I had issues with this function. The cert on my Nexus server is a wild card, so it doesn't have a common name or anything on it that is _specific_ to the server nor that would match the given FQDN that I pass in as a package parameter.

Also, despite my given said FQDN, it ended up using the hostname. My _expectation_ as a user is if I provide an FQDN, then that should be used (regardless of what the cert may say... if I've provided it and I got it wrong, then it's my own fault for shooting myself in the foot (imho)).

Windos commented on automatic/nexus-repository/tools/helpers.ps1:
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
+          $script:OverriddenDomains[$CertDomain]
+        }
+      } else {
+        Write-Warning "Could not figure out SSL configuration for $($env:ComputerName)"
+        $env:ComputerName
```
I would _personally_ prefer this to default to localhost rather than computer name. Some environments (mine 😅) aren't going to resolve the computer name to something useful for the purposes of ensuring Nexus started.

Windos reviewed (commented):
Ok, so, I tested this against a testing restore of a "real" Nexus instance that had an OrientDB. This Nexus instance uses S3 storage to host the configured repos. Despite not having access to this storage, prior to the upgrade Nexus started up fine. After the upgrade, Nexus is now no longer starting up and I can't see a _specific_ reason for this in the logs. I actually think it was to do with a NuGet v2 feed (and so it was a Nexus issue and not an upgrade issue).

Chasing that hunch, I did a take 2 and removed the suspect repo before going through the migration, and it worked!
![image](https://github.com/user-attachments/assets/d4e45738-b158-4781-9854-1103388015a7)

I did run into some other rough edges which I've commented directly on the code itself. The result of those rough edges is that at the end of the DB migration and Nexus upgrade was it couldn't detect that Nexus was back up and running (when it absolutely was.)
![image](https://github.com/user-attachments/assets/3ef3e7a8-6031-4197-8999-88747db9abe7)

tl;dr the migration and upgrade _work_, but in my case the rough edge of "not using the FQDN I provided" resulted in the package install/upgrades to "fail" due to not being able to validate that Nexus web interface was running.

TheCakeIsNaOH reviewed (commented):
I'm not seeing anything glaringly wrong, but as I don't run Nexus on Windows, there may be something I missed.

Windos commented on automatic/nexus-repository/tools/helpers.ps1:
```diff
@@ -102,6 +172,359 @@ function Wait-NexusAvailability {
+    $CurrentVersion = (Get-NexusVersion -ProgramDir $ProgramDir),
+
+    # The only version of Nexus we should be upgrading from
+    $RequiredVersion = "3.70.1",
```
Just one thing, there have been patches beyond this which will still need to be migrated from when going up to 3.71 and beyond. e.g. My Nexus instance is currently 3.70.3.1.

My suggestion would be to make this a Minimum Required Version (so test that the installed version is equal to or above without going into the 3.71.

Windos reviewed (changes requested):
Other than my one comment, this all went fine and all the code looks good to me. I did test this in practice (with a test restore of a real Nexus instance).

pauby commented:
@JPRuskin can you look at the last comment left and update the PR?

JPRuskin committed:
(nexus-repository) Updates Install Method for Latest Changes

Nexus 3.78.0 changed the method for installation, including a new .bat
file instead of simply calling the exe. On investigation, this batch
contains a user prompt with no way to automate it nicely.

After discussion of options including the following:
- Calling the exe with the new install arg
- Stripping the call to the user prompt from the .bat and calling that
- Replicating the bat logic in PowerShell

This commit adds a new Install-NexusRepositoryService function that
replicates the logic contained in the batch file.

AdmiringWorm reviewed (dismissed):
LGTM, however due to lack of knowledge about nexus I would defer to @Windos to give the final approval before we can merge this PR in.
